README.md
# emoji-vote

<p align="center">
  <a href="https://jsfiddle.net/fireyy/q1c2zcya/" target="_blank">
    <img src="demo.png" width="213">
    <br>
    Live Demo
  </a>
</p>

> Emoji vote component.

- Idea from [支付宝商户评论页的满意度 Emoji 会跟随用户的当前选择看向该 Icon](http://uedetail.com/archives/3832).
- Emoji from [EmojiOne™ 2](https://emojitwo.github.io/).

## Install

This project uses [node](http://nodejs.org) and [npm](https://npmjs.com). Go check them out if you don't have them locally installed.

```sh
$ npm install --save emoji-vote
```

Then with a module bundler like [rollup](http://rollupjs.org/) or [webpack](https://webpack.js.org/), use as you would anything else:

```javascript
// using ES6 modules
import emojiVote from 'emoji-vote'

// using CommonJS modules
var emojiVote = require('emoji-vote')
```

The [UMD](https://github.com/umdjs/umd) build is also available on [unpkg](https://unpkg.com):

```html
<script src="//unpkg.com/emoji-vote/dist/emoji-vote.cjs.js"></script>
```

This exposes the `emojiVote()` function as a global.

* * *

## Usage

```html
<div class="emoji">
  <input type="radio" name="emojiVote" id="emojiVote_dissatisfied" value="dissatisfied">
  <label for="emojiVote_dissatisfied">不满</label>
  <input type="radio" name="emojiVote" id="emojiVote_general" value="general">
  <label for="emojiVote_general">一般</label>
  <input type="radio" name="emojiVote" id="emojiVote_satisfied" value="satisfied">
  <label for="emojiVote_satisfied">满意</label>
  <input type="radio" name="emojiVote" id="emojiVote_awesome" value="awesome">
  <label for="emojiVote_awesome">满意</label>
</div>
```

```js
import emojiVote from 'emoji-vote'

emojiVote('.emoji', {
  size: 30
})
```

## Examples & Demos

[**Real Example on JSFiddle**](https://jsfiddle.net/fireyy/q1c2zcya/) ➡️

## API

emoji-vote's API is organized as follows:

### `emojiVote(el: String|Element, options: Object)`

emojiVote will account for the following properties in options:

  * `emoji` The emoji data array, default:

    ```js
    [
      `
        <circle cx="20.5" cy="26.6" r="5"/>
        <circle cx="43.5" cy="26.6" r="5"/>
        <path d="M23 47.6c5.8-4.8 12.2-4.8 18 0 .7.6 1.3-.4.8-1.3-1.8-3.4-5.3-6.5-9.8-6.5s-8.1 3.1-9.8 6.5c-.5.9.1 1.9.8 1.3"/>
      `,
      `
        <circle cx="32" cy="45.1" r="7"/>
        <circle cx="20.2" cy="25" r="4.5"/>
        <circle cx="42.7" cy="25" r="4.5"/>
      `,
      ...
    ]
    ```

  * `render` The render function for create svg, default:

    ```js
      function (data, size) {
        return `
          <svg class="emoji-icon" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 64 64" style="font-size: ${size}px;">
            <circle class="emoji-facial" cx="32" cy="32" r="30" fill="#fff" stroke="#999" stroke-width="3" />
            <g class="emoji-facial-features" fill="#999">
              ${data}
            </g>
          </svg>
        `
      }
    ```

  * `size` The emoji size, default to `20px`.

## TODO

- [x] custom emoji svg template.
- [ ] custom emoji color config


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.3.1"></a>
## [1.3.1](https://github.com/fireyy/emoji-vote/compare/v1.3.0...v1.3.1) (2017-12-12)



<a name="1.3.0"></a>
# [1.3.0](https://github.com/fireyy/emoji-vote/compare/v1.2.0...v1.3.0) (2017-11-30)


### Features

* custom emoji svg template ([aabd902](https://github.com/fireyy/emoji-vote/commit/aabd902))



<a name="1.2.0"></a>
# [1.2.0](https://github.com/fireyy/emoji-vote/compare/v1.1.0...v1.2.0) (2017-11-30)


### Features

* rename to emoji-vote ([959d0f1](https://github.com/fireyy/emoji-vote/commit/959d0f1))



<a name="1.1.0"></a>
# 1.1.0 (2017-11-30)


### Features

* release ([792ff7b](https://github.com/fireyy/emoji-vote/commit/792ff7b))


rollup.config.js
import resolve from 'rollup-plugin-node-resolve';
import commonjs from 'rollup-plugin-commonjs';
import buble from 'rollup-plugin-buble';
import pkg from './package.json';
import sass from 'rollup-plugin-sass';
import autoprefixer from 'autoprefixer';
import postcss from 'postcss';

const config = {
  input: 'src/index.js',
  output: [
    {
      file: pkg.main,
      format: 'umd',
      name: 'emojiVote'
    },
    {
      file: pkg.module,
      format: 'es'
    }
  ],
  plugins: [
    sass({
      processor: css => postcss([autoprefixer])
      .process(css)
      .then(result => result.css),
      insert: true
    }),
    buble(),
    resolve(),
    commonjs()
  ]
}

if (process.env.NODE_ENV === 'development') {
  config.plugins.push(
    require('rollup-plugin-serve')({
      contentBase: __dirname,
      host: 'localhost',
      port: 3000,
      open: false
    }),
    require('rollup-plugin-livereload')()
  )
}

export default config


src/index.js
import './style.scss'

const defaultRender = (data, size) => {
  return `
    <svg class="emoji-icon" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 64 64" style="font-size: ${size}px;">
      <circle class="emoji-facial" cx="32" cy="32" r="30" fill="#fff" stroke="#999" stroke-width="3" />
      <g class="emoji-facial-features" fill="#999">
        ${data}
      </g>
    </svg>
  `
}

const defaultEmojis = [
  `
    <circle cx="20.5" cy="26.6" r="5"/>
    <circle cx="43.5" cy="26.6" r="5"/>
    <path d="M23 47.6c5.8-4.8 12.2-4.8 18 0 .7.6 1.3-.4.8-1.3-1.8-3.4-5.3-6.5-9.8-6.5s-8.1 3.1-9.8 6.5c-.5.9.1 1.9.8 1.3"/>
  `,
  `
    <circle cx="32" cy="45.1" r="7"/>
    <circle cx="20.2" cy="25" r="4.5"/>
    <circle cx="42.7" cy="25" r="4.5"/>
  `,
  `
    <circle cx="20.5" cy="26.6" r="5"/>
    <circle cx="43.5" cy="26.6" r="5"/>
    <path d="M44.6 40.3c-8.1 5.7-17.1 5.6-25.2 0-1-.7-1.8.5-1.2 1.6 2.5 4 7.4 7.7 13.8 7.7s11.3-3.6 13.8-7.7c.6-1.1-.2-2.3-1.2-1.6"/>
  `,
  `
    <path d="M49 38c0-.8-.5-1.8-1.8-2.1-3.5-.7-8.6-1.3-15.2-1.3s-11.7.7-15.2 1.3c-1.3.3-1.8 1.3-1.8 2.1 0 7.3 5.6 14.6 17 14.6S49 45.3 49 38"/>
    <path d="M28.5 26.9c-1.9-5.1-4.7-7.7-7.5-7.7s-5.6 2.6-7.5 7.7c-.2.5.8 1.4 1.3.9 1.8-1.9 4-2.7 6.2-2.7s4.4.8 6.2 2.7c.6.5 1.5-.4 1.3-.9"/>
    <path d="M50.4 26.9c-1.9-5.1-4.7-7.7-7.5-7.7s-5.6 2.6-7.5 7.7c-.2.5.8 1.4 1.3.9 1.8-1.9 4-2.7 6.2-2.7 2.3 0 4.4.8 6.2 2.7.5.5 1.5-.4 1.3-.9"/>
    <path fill="#fff" d="M44.7 38.3c-2.2-.4-6.8-1-12.7-1-5.9 0-10.5.6-12.7 1-1.3.2-1.4.7-1.3 1.5.1.4.1 1 .3 1.6.1.6.3.8 1.3.8h24.9c1 0 1.1-.2 1.3-.8.1-.6.2-1.1.3-1.6 0-.8-.1-1.3-1.4-1.5"/>
  `
]

export default function (el, options = {}) {
  let {
    emojis = defaultEmojis,
    size = 20,
    render = defaultRender
  } = options
  const $el = typeof el === "string" ? document.querySelector(el) : el
  // Create Template
  $el.classList.add("emoji-vote")
  Array.prototype.slice.call($el.querySelectorAll("label")).forEach((label, i) => {
    label.insertAdjacentHTML('afterbegin', render(emojis[i], size))
  })
}
